File: core/src/main/java/com/lomekwi/cave/pipeline/audio/AudioFrameSink.java
package com.lomekwi.cave.pipeline.audio;

import com.google.common.eventbus.Subscribe;
import com.lomekwi.cave.app.App;
import com.lomekwi.cave.project.ProjectBackgroundedEvent;
import com.lomekwi.cave.project.ProjectFrontedEvent;
import com.lomekwi.cave.resource.decoder.AudDecRes;

import java.util.Arrays;
import java.util.concurrent.Future;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;

public class AudioFrameSink {
    private volatile AudioFrameMixer afm = new AudioFrameMixer();
    private Future<?> currentFuture;

    @Subscribe
    public void sink(AudFrame frame) {
        frame.track.getWorker().getSinkPhaser().register();
        afm.submit(frame);
    }

    @Subscribe
    public void onProjectFronted(ProjectFrontedEvent event) {
        stopMixer();
        afm = new AudioFrameMixer();
        currentFuture = App.workerExecutor.submit(afm);
    }

    @Subscribe
    public void onProjectBackgrounded(ProjectBackgroundedEvent event) {
        stopMixer();
    }

    private void stopMixer() {
        Future<?> f = currentFuture;
        currentFuture = null;
        if (f == null) return;
        afm.stop();
        try {
            f.get(200, TimeUnit.MILLISECONDS);
        } catch (Exception e) {
            f.cancel(true);
        }
    }

    protected static class AudioFrameMixer implements Runnable {
        private static final long POLL_TIMEOUT_MILLIS = 20;

        private final LinkedBlockingQueue<AudFrame> frames = new LinkedBlockingQueue<>();
        private final float[] output = new float[AudDecRes.FRAME_SIZE];
        private volatile boolean stopped;

        public void stop() {
            stopped = true;
        }

        @Override
        public void run() {
            while (!stopped) {
                Arrays.fill(output, 0f);
                AudFrame f;
                try {
                    f = frames.poll(POLL_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
                } catch (InterruptedException e) {
                    break;
                }
                if (f == null) continue;
                do {
                    int j = 0;
                    for (var sample : f.getSamples()) {
                        output[j] += sample;
                        j++;
                    }
                    f.track.getWorker().getSinkPhaser().arriveAndDeregister();
                    if (stopped || Thread.currentThread().isInterrupted()) break;
                } while ((f = frames.poll()) != null);
                if (stopped || Thread.currentThread().isInterrupted()) break;
                clamp(output);
                App.audioOut.writeSamples(output);
            }
        }

        private void clamp(float[] samples) {
            for (int i = 0; i < samples.length; i++) {
                samples[i] = Math.max(-1.0f, Math.min(1.0f, samples[i]));
            }
        }

        public void submit(AudFrame f) {
            frames.add(f);
        }
    }
}

File: core/src/main/scala/com/lomekwi/cave/pipeline/audio/AudioFrameSink.scala
package com.lomekwi.cave.pipeline.audio

import com.google.common.eventbus.Subscribe
import com.lomekwi.cave.app.App
import com.lomekwi.cave.project.ProjectBackgroundedEvent
import com.lomekwi.cave.project.ProjectFrontedEvent
import com.lomekwi.cave.resource.decoder.AudDecRes

import java.util
import java.util.concurrent.{Future, LinkedBlockingQueue, TimeUnit}
import scala.compiletime.uninitialized

class AudioFrameSink {
  @volatile private var afm: AudioFrameSink.AudioFrameMixer = new AudioFrameSink.AudioFrameMixer()
  private var currentFuture: Future[?] = uninitialized

  @Subscribe
  def sink(frame: AudFrame): Unit = {
    frame.track.getWorker.getSinkPhaser.register()
    afm.submit(frame)
  }

  @Subscribe
  def onProjectFronted(event: ProjectFrontedEvent): Unit = {
    stopMixer()
    afm = new AudioFrameSink.AudioFrameMixer()
    currentFuture = App.workerExecutor.submit(afm)
  }

  @Subscribe
  def onProjectBackgrounded(event: ProjectBackgroundedEvent): Unit = {
    stopMixer()
  }

  private def stopMixer(): Unit = {
    val f = currentFuture
    currentFuture = null
    if (f != null) {
      afm.stop()
      try {
        f.get(200, TimeUnit.MILLISECONDS)
      } catch {
        case _: Exception =>
          f.cancel(true)
      }
    }
  }
}

object AudioFrameSink {
  private[audio] class AudioFrameMixer extends Runnable {
    private final val frames: LinkedBlockingQueue[AudFrame] = new LinkedBlockingQueue[AudFrame]()
    private final val output: Array[Float] = new Array[Float](AudDecRes.FRAME_SIZE)
    @volatile private var stopped: Boolean = false

    def stop(): Unit = {
      stopped = true
    }

    override def run(): Unit = {
      while (!stopped) {
        util.Arrays.fill(output, 0f)
        var f: AudFrame = null
        try {
          f = frames.poll(AudioFrameMixer.POLL_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS)
        } catch {
          case _: InterruptedException =>
            return
        }
        if (f != null) {
          var continueLoop = true
          while (continueLoop) {
            var j = 0
            for (sample <- f.getSamples) {
              output(j) += sample
              j += 1
            }
            f.track.getWorker.getSinkPhaser.arriveAndDeregister()
            if (stopped || Thread.currentThread().isInterrupted) {
              continueLoop = false
            } else {
              f = frames.poll()
              continueLoop = f != null
            }
          }
          if (stopped || Thread.currentThread().isInterrupted) {
            return
          }
          clamp(output)
          App.audioOut.writeSamples(output)
        }
      }
    }

    private def clamp(samples: Array[Float]): Unit = {
      var i = 0
      while (i < samples.length) {
        samples(i) = Math.max(-1.0f, Math.min(1.0f, samples(i)))
        i += 1
      }
    }

    def submit(f: AudFrame): Unit = {
      frames.add(f)
    }
  }

  private object AudioFrameMixer {
    private final val POLL_TIMEOUT_MILLIS = 20L
  }
}
